nemo_skills/inference/reward_model.py

```diff
@@ -0,0 +1,204 @@
+# Copyright (c) 2024, NVIDIA CORPORATION.  All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import importlib
+import json
+import logging
+import sys
+from copy import deepcopy
+from dataclasses import asdict, field
+from pathlib import Path
+
+import hydra
+from tqdm import tqdm
+
+from nemo_skills.code_execution.sandbox import get_sandbox, sandbox_params
+from nemo_skills.inference.server.code_execution_model import ErrorRecoveryConfig, server_params
+from nemo_skills.inference.server.reward_model import RequestException, get_reward_model
+from nemo_skills.prompt.utils import get_prompt
+from nemo_skills.utils import get_fields_docstring, get_help_message, nested_dataclass, setup_logging
+
+LOG = logging.getLogger(__file__)
+
+
+@nested_dataclass(kw_only=True)
+class RewardModelConfig:
+    """LLM reward model parameters."""
+
+    output_file: str | None = None  # Where to save the generations
+    # Inference server configuration {server_params} {error_recovery_params}
+    server: dict = field(default_factory=dict)
+    # Sandbox configuration {sandbox_params}
+    sandbox: dict = field(default_factory=dict)
+    # Prompt configuration - path to yaml files
+    prompt_template: str | None = None  # not required for OpenAI server
+    prompt_config: str | None = None  # we will fetch it from dataset dir if not provided
+    examples_type: str | None = None  # to be able to customize few-shot examples
+
+    # Can specify one of the existing datasets.
+    dataset: str | None = None
+    split: str | None = None  # Can be train, validation, test or train_full (train + validation)
+    input_file: str | None = None  # Can directly specify an input file, if using a custom dataset
+    input_dir: str | None = None  # Can specify an input direct
```

Comment:
Can you please add more details in the comment on how this is used exactly?

Reply:
After seeing how it's used, I'd suggest to remove input_dir and random_seed as well. Instead (unless it's not convenient for some reason I don't see yet), let's use a more general parameter `input_files` that we e.g. use here https://github.com/Kipok/NeMo-Skills/blob/main/nemo_skills/inference/llm_math_judge.py#L43 (and in a few other places as well)

Reply:
And similarly how we do for llm math judge, we can probably include the scores directly inside the original input files

Reply:
And I guess if we do all that, then it makes sense to share reward model and llm_math_judge pipeline scripts instead of the generate.py. I guess those are almost the same in principle (both are used to "score" existing generations) and whatever changes we do to one of them, we should do to another one. Sorry that this is not what we originally discussed, but seems to make more sense to me now - let me know if you don't think that's the right way to go about that

Reply:
Maybe it'll help if we outline some use-cases and user stories. In my estimation there were a few things we wanted to support.

## Use Cases
1. A user wants to score a directory of generations from `ns generate`
2. A user wants to use a reward model to score/filter training data
3. A user wants to evaluate multiple reward models to compare their scores on the same data.


## User Stories

1. A user must be able to score a single file [list] (arbitrary source) with a single reward model
2. A user must be able to score a full directory of outputs from `ns generate`  (e.g., `n` random seeds, or `greedy` generation) with a single reward model
3. A user must be able to score a single file [list] (arbitrary source) with several reward models
4. A user must be able to score a full directory of outputs from `ns generate`  (e.g., `n` random seeds, or `greedy` generation) with several reward models
5. A user may share the results of their `ns generate` command with another user. Both users may want to score these files in different ways at any point in time.

Open to feedback on these use-cases and user-stories.

## PR feedback

With this in mind, let's look at some of the feedback.

> I'd suggest to remove input_dir and random_seed

These are useful because it helps us support the output of `ns generate` fairly seamlessly.


> let's use a more general parameter `input_files`

I think this is undesirable to have this be the _sole_ way to specify input files if you want to score, e.g., 256 generations with different random seeds. It would be way too cumbersome to specify all files individually.

What I think might be reasonable is to convert the `input_file` parameter I've written to `input_files`, but keep the input directory/seed parameters though.

> we can probably include the scores directly inside the original input files

The issue I see with this, is this is a potentially destructive operation--namely, if there is already a reward key, you could overwrite the data there. This is detrimental if you are (1) trying to score with different reward models, and/or (2) you are writing to a shared location like (solutions in `igitman` directories), this could easily lead to data corruption (the different samples end up being scored by different models, but saved to the same key for their respective samples within the same file) and write contention (multiple people trying to write to the same file at the same time, possibly with different data).  Some of this could be partially mitigated by allowing the user to specify the output key (this is a change I should make already) and checking for the existence of the key before writing, but I think it is safer to make the user specify the directory/file they want it output to, and possibly indicate whether overwriting the data there is okay. Of course they could always specify the input directory to opt in to these risks, but by default it is a little safer.

> if we do all that, then it makes sense to share reward model and llm_math_judge pipeline scripts instead of the generate.py

It looks like the only differences between these pipelines (since most args are/can be passed through ctx/extra_arguments) is effectively that `random_seed` and `dependent_jobs` are not supported for llm_math_judge. I personally think those are useful features here based on the outlined use-cases/user-stories and other discussion here, but feel free to provide additional context.

Reply:
input_files support glob patterns, so you can just use something like `output_dir/output-rs*.jsonl` to support typical generations with random seeds. Moreover, you can even use multiple patterns like this, e.g. for running on top of eval data you can say `output_dir/math/output*.json output_dir/aime24/output*.jsonl` or even something like `output_dir/**/ouptut*.jsonl` to pick all benchmarks, so I think it's strictly more flexible and convenient than output_dir/random seeds.

About data corruption, that's certainly a valid concern and something I am personally undecided about. The problem with output_dir is that if you start specifying patterns like above, it's becoming unclear how we should move those input files and so you would have to split the command into many pieces and then manage data in different locations. What do you think about instead we check if the requested output key is already present in any of the input files and then we just raise an error and ask a user to explicitly specify some parameter to force override? That should solve the problem of corrupting data while also keeping the pipeline simpler I think

Reply:
> What do you think about instead we check if the requested output key is already present in any of the input files and then we just raise an error and ask a user to explicitly specify some parameter to force override? That should solve the problem of corrupting data while also keeping the pipeline simpler I think

This is still unsafe in the case where multiple users (or one user, multiple concurrent jobs) are scoring the same file at the same time. I think this is a fairly valid use case because you may want to queue multiple scoring jobs at the same time whenever you're logged onto the cluster and just wait for their result later.

Assume you have two concurrent jobs to score a file with a NeMo trained RM (`NeMo-RM`) and the QWEN RM (`QWEN-RM`). `NeMo-RM` and `QWEN-RM` could open the same generation file (`G`) and start scoring it if both jobs check for output keys before the other job finished. If it runs like `inference/llm_math_judge.py` does, both jobs will start appending to the same tempfile. Depending on whether the tempfile is deleted/moved after the first job completes or not, the final output file after both jobs have finished (`G'`)  will have somewhere between 0 to `2X` entries where `X` is the number of scored samples and there is no guaranteed correspondence between the index in the original input and the index in the new output. If the jobs shared an output key, it will be indiscernible which version of the score in `G'` came from `NeMo-RM` and which one came from `QWEN-RM`. Or there could be no version of a sample in the final scored version at all.

If they do have different output keys (which again, cannot be guaranteed unless you explicitly set it) AND there are no destructive operations until all scoring is finished, you would still have to do some post-processing to merge the entries back together in order to end up with `X` entries and have both scores available in the same line.

Furthermore, if you have `skip_filled` set and/or dependent jobs, you risk further skewing your original dataset by either re-running samples by the same model, or skipping samples, dependent on the conditions under which it happens because they do math based on the number of samples in `G` and `G'`.

While the "different output key, shared output to the input file" solution can help prevent the case where you are unsure which model the score originated from, no mechanism exists to avoid the issues that come with multiple writers to the same file. The "ability to specify output file" however, does give you a mechanism to avoid it where you can save to separate directories where you can guarantee the integrity of the data and separately later combine all of outputs back into one file. In my opinion, I would rather have an option to avoid these issues with multiple writers than gain a slightly more ergonomic pipeline syntax.

Reply:
ok, that makes sense. But we still need a way to easily support another user story which is rescoring the output of `ns eval` that doesn't only give you multiple seeds, but also potentially multiple benchmark folders. What do you think about having `input_dir` parameter and then we would just search for anything matching `{input_dir}/**/output*.jsonl` and run RM on that and then copy the full folder structure to `output_dir`? And maybe we can give an ability to customize that search regex in case someone only wants to rescore greedy results a subset of random seeds

Reply:
and I guess ideally we would also support `num_jobs` argument similar to how `ns eval` does it and would parallelize the RM scoring across that number of nodes. But maybe we can leave hat for later since it requires using an ssh tunnel to read how many files we have in that regex on cluster and then properly format the arguments for each individual job. I guess for now we can just assume that it all fits within a single job, which is a reasonable assumption in most cases I think given that RM is a lot faster than generation itself

Reply:
What's the order of magnitude of "multiple benchmark folders"? 5, 10, 100? If it's less than 10 I wouldn't even worry about making multiple calls to `ns generate` for different benchmarks. If you expect it to grow far past that, yeah I'm sure we can come up with some interface like what you just mentioned.

Reply:
less than 10 in almost all cases. But it's still convenient to have any easy way to run RM on "everything" instead of specifying exact random seeds used, so I'd still recommend going for regex instead of an explicit random_seed parameter. We almost always would want to combine multiple random seeds into a single job (well, at least as soon as we support vllm/trtllm, for nemo we might need to parallelize things more)

Reply:
or if you think the current way is better, I guess I'm fine with that as well. We can merge it like this and then see if it's convenient or not and refine the arguments later

Reply:
after thinking a bit more, I think we should indeed keep your current way. And if we need to merge multiple generations together, it's better to add some general functionality inside pipeline/generate.py for that, since it's useful not only for RM, but also for other generation tasks, I guess. And then for benchmarks scoring, we can just have a loop in python that calls generate on each benchmark separately, so it's not a problem (except fusing multiple generations into a single node, but that's fine I think). And again, if it turns out not to be convenient for some reason, we can revise this later

Reply:
I was just warming up to `input_dir`, `input_pattern` and `output_dir`, with folder structure copied from matches within `input_dir` 😆 

But yeah I agree with this result--let's keep it for now and update if its not sufficient.